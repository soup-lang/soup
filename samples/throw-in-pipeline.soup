x: [0, 1, 2, 3]

try
    x |:
        if $ % 4 = 3
            throw 'sync pipeline throw:' + $
catch e
    console.log(e)

try
    x |:
        setTimeout(%, 1)
        if $ % 4 = 3
            throw 'async pipeline throw:' + $
catch e
    console.log(e)

func aGet(n, %%)
    if n % 3 = 2
        throw 'aGet throw:' + n

func sGet(n)
    if n % 3 = 2
        throw 'sGet throw:' + n

func piping(%%)
    try
        x |: aGet($, %%)
    catch e
        console.log(e)
    try
        x |: sGet($)
    catch e
        console.log(e)

piping(%%)

console.log('======')

y: [
    [0, 1],
    [2, 3],
]

func mixing_pipes(%%)
    try
        y |:
            console.log('#x', $.join(':'))
            $ |:
                aGet($, %%)
    catch e
        console.log('%x', e)

    try
        y |:
            console.log('#y', $.join(':'))
            $ |:
                sGet($)
    catch e
        console.log('%y', e)

    y |:
        console.log('#z', $.join(':'))
        $ |:
            aGet($, %%)

try
    mixing_pipes(%%)
catch e
    console.log('%z', e)

console.log('end')
