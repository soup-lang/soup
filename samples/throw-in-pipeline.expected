sync pipeline throw:3
async pipeline throw:3
aGet throw:2
sGet throw:2
======
#x 0:1
#x 2:3
%x aGet throw:2
#y 0:1
#y 2:3
%y sGet throw:2
#z 0:1
#z 2:3
%z aGet throw:2
end
